src/cdomains/unionFind.ml

```diff
@@ -0,0 +1,743 @@
+(**
+   The Union Find is used by the C-2PO Analysis.
+   This file contains the code for a quantitative union find and the quantitative finite automata.
+   They will be necessary in order to construct the congruence closure of terms.
+*)
+open Batteries
+open GoblintCil
+open DuplicateVars
+module M = Messages
+
+exception Unsat
+
+type ('v, 't) term = Addr of 'v | Aux of 'v * 't | Deref of ('v, 't) term * Z.t * 't [@@deriving eq, ord, hash]
+type ('v, 't) prop = Equal of ('v, 't) term * ('v, 't) term * Z.t | Nequal of ('v, 't) term * ('v, 't) term * Z.t
+                   | BlNequal of ('v, 't) term * ('v, 't) term
+[@@deriving eq, ord, hash]
+
+(** The terms consist of address constants and dereferencing function with sum of an integer.
+    The dereferencing function is parametrized by the size of the element in the memory.
+    We store the CIL expression of the term in the data type, such that it it easier to find the types of the dereferenced elements.
+    Also so that we can easily convert back from term to Cil expression.
+*)
+module T = struct
+  type exp = Cil.exp
+
+  let bitsSizeOfPtr () = Z.of_int @@ bitsSizeOf (TPtr (TVoid [],[]))
+
+  (* equality of terms should not depend on the expression *)
+  let compare_exp _ _ = 0
+  let equal_exp _ _ = true
+  let hash_exp _ = 1
+
+  (* we store the varinfo and the Cil expression corresponding to the term in the data type *)
+  type t = (Var.t, exp[@compare.ignore][@eq.ignore][@hash.ignore]) term [@@deriving eq, ord, hash]
+  type v_prop = (Var.t, exp[@hash.ignore]) prop [@@deriving hash]
```

Comment:
Unfortunately the conversion back to expression is a bit complicated, because I convert arrays like `a[3]` to pointer terms `*(a+3)`. This way I can also express properties about arrays. And I do the same with structs, I calculate the offset of structs by using the Evalnt query on the cil offset. Also, there might be casts involved in the expressions, so in practice it's not always straightforward how to reconvert a term back to the cil expression. 
Given that we only analyze terms that are pointers or 64-bit integers, in my opinion it's not a problem to merge equivalent values with different expressions. For example `(int*)pointer1` and `(long*)pointer1` still have the same value, even if they don't have the same type. 
The difference in the analysis is that the dereferenced term of `(int*)pointer1` is ignored, as it is not 64-bits. While if we dereference `(long*)pointer1` it will be added to the domain.